internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,156 @@
+		return nil, fmt.Errorf(ZeroSizeBufferError)
+	}
+	b := &InMemoryWriteBuffer{
+		bufferSize: int64(bufferSizeMB * MiB),
+	}
 	// TODO: set mtime attribute.
-	return b
+	logger.Debugf("TODO: set mtime attribute.")
+	return b, nil
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+// Allocates memory to currentBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureCurrentBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
 	}
+	if b.current.buffer == nil {
+		b.current.buffer = make([]byte, 0, b.bufferSize)
+		b.current.offset.end = b.bufferSize
+	}
+	return nil
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+// Allocates memory to flushedBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureFlushedBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.flushed.buffer == nil {
+		b.flushed.buffer = make([]byte, 0, b.bufferSize)
+	}
 	return nil
 }
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := offset + dataSize
+	if dataSize == 0 {
+		return nil
+	}
+
+	// Ensure b.currentBuffer != nil
+	if err := b.ensureCurrentBuffer(); err != nil {
+		return err
+	}
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if offset < b.current.offset.start ||
+		offset > b.current.offset.end {
+		// TODO: finalise write and trigger temp file flow.
+		logger.Debugf("TODO: finalise write and trigger temp file flow.")
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written to the current buffer, write it.
+	if receivedContentEndOffset <= b.current.offset.end {
+		return b.copyDataToBuffer(offset, receivedContentEndOffset, content)
+	}
+	// Else, write the chunk of data that can be written to the current buffer block.
+	l := b.current.offset.end - offset
+	err := b.copyDataToBuffer(offset, b.current.offset.end, content[:l])
+	if err != nil {
+		return err
+	}
+	//TODO: get status of last write to GCS. If unsuccessful, trigger temp file flow.
+	logger.Debugf("TODO: get status of last write to GCS and if " +
+		"unsuccessful, trigger temp file flow.")
+	if err = b.advanceToNextChunk(); err != nil {
+		return err
+	}
+	// TODO: trigger async upload to GCS for flushedBuffer.
+	logger.Debugf("TODO: trigger async upload to GCS for flushedBuffer.")
+
+	// Write the remaining data to currentBuffer.
+	return b.copyDataToBuffer(b.current.offset.start, receivedContentEndOffset, content[l:])
+}
+
+// Copies received content to currentChunk if it has capacity.
+func (b *InMemoryWriteBuffer) copyDataToBuffer(contentStartOffset, contentEndOffset int64, content []byte) error {
```

Comment:
contentEndOffset is redundant. It's same as `contentStartOffset + len(content)` .